File: src/main/java/io/github/tanguygab/conditionalactions/ConditionalActions.java
package io.github.tanguygab.conditionalactions;

import io.github.tanguygab.conditionalactions.actions.ActionManager;
import io.github.tanguygab.conditionalactions.commands.*;
import io.github.tanguygab.conditionalactions.conditions.ConditionManager;
import io.github.tanguygab.conditionalactions.customcommands.CustomCommandManager;
import io.github.tanguygab.conditionalactions.hooks.papi.CAExpansion;
import io.github.tanguygab.conditionalactions.hooks.papi.PAPIExpansion;
import lombok.Getter;
import org.bukkit.OfflinePlayer;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.event.HandlerList;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;

import java.util.*;

public final class ConditionalActions extends JavaPlugin {

    @Getter private static ConditionalActions instance;
    private final Map<String, CACommand> subcommands = new HashMap<>();
    private final List<PAPIExpansion> expansions = new ArrayList<>();

    @Getter private DataManager dataManager;
    @Getter private ActionManager actionManager;
    @Getter private ConditionManager conditionManager;
    @Getter private CustomCommandManager customCommandManager;


    @Override
    public void onEnable() {
        instance = this;
        saveDefaultConfig();
        reloadConfig();

        dataManager = new DataManager();
        conditionManager = new ConditionManager(this);
        customCommandManager = new CustomCommandManager(this);
        actionManager = new ActionManager(this,getConfig().getString("argument-separator",","));

        expansions.add(new CAExpansion(this));
        expansions.add(new PAPIExpansion(this,"condition") {
            @Override
            public Object parse(OfflinePlayer player, String params) {
                return plugin.getConditionManager().getCondition(params).isMet(player);
            }
        });
        expansions.forEach(PAPIExpansion::register);

        subcommands.put("reload",new ReloadCommand(this));
        subcommands.put("execute",new ExecuteCommand(this));
        subcommands.put("group",new GroupCommand(this));
        subcommands.put("condition",new ConditionCommand(this));
    }

    @Override
    public void onDisable() {
        HandlerList.unregisterAll(this);
        try {
            getServer().getAsyncScheduler().cancelTasks(this);
            getServer().getGlobalRegionScheduler().cancelTasks(this);
        } catch (NoSuchMethodError e) {
            getServer().getScheduler().cancelTasks(this);
        }

        subcommands.clear();

        expansions.forEach(PAPIExpansion::unregister);
        expansions.clear();
    }

    public void async(Runnable run) {
        try {
            getServer().getAsyncScheduler().runNow(this, task -> run.run());
        } catch (NoSuchMethodError e) {
            getServer().getScheduler().runTaskAsynchronously(this, run);
        }
    }

    public void sync(Player player, Runnable run) {
        try {
            if (player == null) getServer().getGlobalRegionScheduler().run(this, task -> run.run());
            else player.getScheduler().run(this, task -> run.run(), null);
        } catch (NoSuchMethodError e) {
            getServer().getScheduler().runTask(this, run);
        }
    }

    @Override
    public boolean onCommand(@NotNull CommandSender sender, @NotNull Command command, @NotNull String label, String[] args) {
        String arg = args.length > 0 ? args[0] : "";
        if (subcommands.containsKey(arg)) {
            args = Arrays.copyOfRange(args,1, args.length);
            subcommands.get(arg).onCommand(sender,args);
            return true;
        }
        return true;
    }

    @Override
    public List<String> onTabComplete(@NotNull CommandSender sender, @NotNull Command command, @NotNull String alias, String[] args) {
        String arg = args.length > 0 ? args[0] : "";
        if (!subcommands.containsKey(arg)) return List.of("execute","group","condition","reload");
        args = Arrays.copyOfRange(args,1, args.length);
        return subcommands.get(arg).onTabComplete(sender,args);
    }

    public static String parseReplacements(String string, Map<String, String> replacements) {
        for (String replacement : replacements.keySet())
            string = string.replace(replacement, replacements.get(replacement));
        return string;
    }
}

File: src/main/kotlin/io/github/tanguygab/conditionalactions/ConditionalActions.kt
package io.github.tanguygab.conditionalactions

import com.google.common.io.ByteStreams
import io.github.tanguygab.conditionalactions.actions.ActionManager
import io.github.tanguygab.conditionalactions.commands.CACommands
import io.github.tanguygab.conditionalactions.conditions.ConditionManager
import io.github.tanguygab.conditionalactions.customcommands.CustomCommandManager
import io.github.tanguygab.conditionalactions.hook.papi.CAExpansion
import io.github.tanguygab.conditionalactions.hook.papi.PAPIExpansion
import io.github.tanguygab.conditionalactions.listener.PlayerJoinListener
import io.papermc.paper.plugin.lifecycle.event.types.LifecycleEvents
import net.kyori.adventure.text.minimessage.MiniMessage
import org.bukkit.entity.Player
import org.bukkit.event.HandlerList
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.plugin.messaging.PluginMessageListener
import java.io.File

class ConditionalActions : JavaPlugin(), PluginMessageListener {
    lateinit var expansion: PAPIExpansion
    val servers = mutableMapOf<String, Int>()
    var serverName = "Loading..."

    val mm = MiniMessage.miniMessage()
    lateinit var messages: MessagesFile

    lateinit var dataManager: DataManager
    lateinit var actionManager: ActionManager
    lateinit var conditionManager: ConditionManager
    lateinit var customCommandManager: CustomCommandManager

    override fun onLoad() {
        lifecycleManager.registerEventHandler(LifecycleEvents.COMMANDS) {
            it.registrar().register(CACommands(this).main, listOf("ca"))
        }
    }

    override fun onEnable() {
        INSTANCE = this
        saveDefaultConfig()
        reloadConfig()
        val messagesFiles = File(dataFolder, "messages.yml")
        if (!messagesFiles.exists()) saveResource("messages.yml", true)
        messages = MessagesFile(messagesFiles, mm)

        dataManager = DataManager()
        conditionManager = ConditionManager(this)
        customCommandManager = CustomCommandManager(this)
        actionManager = ActionManager(this, config.getString("argument-separator") ?: ",")

        expansion = CAExpansion(this).also { it.register() }

        server.messenger.registerOutgoingPluginChannel(this, "BungeeCord")
        server.messenger.registerOutgoingPluginChannel(this, CHANNEL)
        server.messenger.registerIncomingPluginChannel(this, CHANNEL, this)
        server.sendPluginMessage(this, CHANNEL, ByteStreams.newDataOutput().apply { writeUTF("start") }.toByteArray())
        server.pluginManager.registerEvents(PlayerJoinListener(this), this)
    }

    override fun onDisable() {
        HandlerList.unregisterAll(this)
        server.asyncScheduler.cancelTasks(this)
        server.globalRegionScheduler.cancelTasks(this)

        server.messenger.unregisterOutgoingPluginChannel(this)
        server.messenger.unregisterIncomingPluginChannel(this)

        servers.clear()
        expansion.unregister()
    }

    fun async(run: Runnable) {
        server.asyncScheduler.runNow(this) { run.run() }
    }

    fun sync(player: Player?, run: Runnable) {
        if (player == null) server.globalRegionScheduler.run(this) { run.run() }
        else player.scheduler.run(this, { run.run() }, null)
    }

    override fun onPluginMessageReceived(channel: String, player: Player, message: ByteArray) {
        if (channel != CHANNEL) return
        ByteStreams.newDataInput(message).apply {
            when (readUTF()) {
                "name" -> serverName = readUTF()
                "update" -> {
                    if (readUTF() != "update") return
                    val size = readInt()
                    (0..<size).forEach { _ ->
                        val info = readUTF().split("|")
                        val server = info[0]
                        if (info[1] == "false")
                            servers.remove(server)
                        else servers[server] = info[2].toInt()
                    }
                }
            }
        }
    }

    companion object {
        lateinit var INSTANCE: ConditionalActions
        const val CHANNEL = "conditionalactions:channel"
    }
}
